src/uptane/tufrepository.cc

```diff
@@ -38,207 +49,95 @@ Json::Value TufRepository::getJSON(const std::string& role) {
-                                         boost::posix_time::duration_from_string(date.substr(11, 8)));
-    return boost::posix_time::second_clock::local_time() > parsed_date;
-  } catch (std::exception e) {
-    throw Uptane::Exception(name_, std::string("Wrong expires datetime field, what(): ") + e.what());
-  }
-  return false;
-}
-
-void TufRepository::verifyRole(const Json::Value& tuf_signed) {
-  std::string role = boost::algorithm::to_lower_copy(tuf_signed["signed"]["_type"].asString());
-  if (!tuf_signed["signatures"].size()) {
-    throw SecurityException(name_, "Missing signatures, verification failed");
-  } else if (tuf_signed["signatures"].size() < thresholds_[role]) {
-    throw UnmetThreshold(name_, role);
-  }
-
-  std::string canonical = Json::FastWriter().write(tuf_signed["signed"]);
-  for (Json::ValueIterator it = tuf_signed["signatures"].begin(); it != tuf_signed["signatures"].end(); it++) {
-    std::string method((*it)["method"].asString());
-    std::transform(method.begin(), method.end(), method.begin(), ::tolower);
-
-    if (method != "rsassa-pss" && method != "rsassa-pss-sha256" && method != "ed25519") {
-      throw SecurityException(name_, std::string("Unsupported sign method: ") + (*it)["method"].asString());
-    }
-    std::string keyid = (*it)["keyid"].asString();
-    if (!keys_.count(keyid)) {
-      throw SecurityException(name_, std::string("Couldn't find a key: ") + keyid);
-    }
-    if (!Crypto::VerifySignature(keys_[keyid], (*it)["sig"].asString(), canonical)) {
-      throw SecurityException(name_, "Invalid signature, verification failed");
-    }
-  }
-  if (hasExpired(tuf_signed["signed"]["expires"].asString())) {
-    throw ExpiredMetadata(name_, role);
-  }
-}
-
-void TufRepository::saveRole(const Json::Value& content) {
-  std::string role = boost::algorithm::to_lower_copy(content["signed"]["_type"].asString());
-  std::ofstream file((path_ / (role + ".json")).string().c_str());
+/**
+ * Note this doesn't check the validity of version numbers, only that the signed part is actually signed.
+ * @param role
+ * @return The contents of the "signed" section
+ */
+Json::Value TufRepository::fetchAndCheckRole(Uptane::Role role, Version version) {
+  Json::Value content = getJSON(version.RoleFileName(role));
+  TimeStamp now(TimeStamp::Now());
+  Json::Value result = root_.UnpackSignedObject(now, name_, role, content);
+  if (role == kRoot) {
+    // Also check that the new root is suitably self-signed
+    Root new_root = Root(name_, result);
+    new_root.UnpackSignedObject(TimeStamp::Now(), name_, kRoot, content);
+  }
+  // UnpackSignedObject throws on error--write the (now known to be valid) content to disk
+  std::ofstream file((path_ / (StringFromRole(role) + ".json")).string().c_str());
   file << content;
   file.close();
+  return result;
+}
+
+void TufRepository::verifyRole(Uptane::Role role, const Json::Value& tuf_signed) {
+  Uptane::TimeStamp now("2017-01-01T01:00:00Z");
```

Comment:
This is only used in testing at the moment--I'd prefer test code didn't depend on the current time.  verifyRole is maybe not a great name though.  I'll try and think of a better one